File: apple/Sources/Autograph/Channel.swift
import Clibautograph
import Foundation

public class Channel {
    var ourIdentityKeyPair: [UInt8]
    var ourSessionKeyPair: [UInt8]
    var theirIdentityKey: [UInt8]
    var theirSessionKey: [UInt8]
    var transcript: [UInt8]
    var sendingKey: [UInt8]
    var receivingKey: [UInt8]
    var sendingNonce: [UInt8]
    var receivingNonce: [UInt8]
    var skippedIndexes: [UInt8]

    public init() {
        ourIdentityKeyPair = createIdentityKeyPair()
        ourSessionKeyPair = createSessionKeyPair()
        theirIdentityKey = createIdentityPublicKey()
        theirSessionKey = createSessionPublicKey()
        transcript = createTranscript()
        sendingKey = createSecretKey()
        receivingKey = createSecretKey()
        sendingNonce = createNonce()
        receivingNonce = createNonce()
        skippedIndexes = createSkippedIndexes(nil)
    }

    public func useKeyPairs(
        _ ourIdentityKeyPair: [UInt8],
        _ ourSessionKeyPair: [UInt8]
    ) -> ([UInt8], [UInt8]) {
        autograph_use_key_pairs(
            &self.ourIdentityKeyPair,
            &self.ourSessionKeyPair,
            ourIdentityKeyPair,
            ourSessionKeyPair
        )
        return getPublicKeys(ourIdentityKeyPair, ourSessionKeyPair)
    }

    public func usePublicKeys(
        _ theirIdentityKey: [UInt8],
        _ theirSessionKey: [UInt8]
    ) {
        autograph_use_public_keys(
            &self.theirIdentityKey,
            &self.theirSessionKey,
            theirIdentityKey,
            theirSessionKey
        )
    }

    public func authenticate(
        _ ourId: [UInt8],
        _ theirId: [UInt8]
    ) throws -> [UInt8] {
        try Autograph.authenticate(
            ourIdentityKeyPair,
            ourId,
            theirIdentityKey,
            theirId
        )
    }

    public func certify(_ data: [UInt8]?) throws -> [UInt8] {
        try Autograph.certify(
            ourIdentityKeyPair,
            theirIdentityKey,
            data
        )
    }

    public func verify(
        _ certifierIdentityKey: [UInt8],
        _ signature: [UInt8],
        _ data: [UInt8]?
    ) -> Bool {
        Autograph.verify(
            theirIdentityKey,
            certifierIdentityKey,
            signature,
            data
        )
    }

    public func keyExchange(_ isInitiator: Bool) throws -> [UInt8] {
        let (
            transcript,
            ourSignature,
            sendingKey,
            receivingKey
        ) = try Autograph.keyExchange(
            isInitiator,
            ourIdentityKeyPair,
            ourSessionKeyPair,
            theirIdentityKey,
            theirSessionKey
        )
        self.transcript = transcript
        self.sendingKey = sendingKey
        self.receivingKey = receivingKey
        return ourSignature
    }

    public func verifyKeyExchange(_ theirSignature: [UInt8]) throws {
        try Autograph.verifyKeyExchange(
            transcript,
            ourIdentityKeyPair,
            theirIdentityKey,
            theirSignature
        )
    }

    public func encrypt(_ plaintext: [UInt8]) throws -> (UInt32, [UInt8]) {
        try Autograph.encrypt(
            sendingKey,
            &sendingNonce,
            plaintext
        )
    }

    public func decrypt(_ ciphertext: [UInt8]) throws -> (UInt32, [UInt8]) {
        try Autograph.decrypt(
            receivingKey,
            &receivingNonce,
            &skippedIndexes,
            ciphertext
        )
    }
}

File: android/src/main/java/sh/autograph/Channel.kt
package sh.autograph

public class Channel {
    private var ourIdentityKeyPair: ByteArray = KeyPair.createIdentityKeyPair()
    private var ourSessionKeyPair: ByteArray = KeyPair.createSessionKeyPair()
    private var theirIdentityKey: ByteArray = KeyPair.createIdentityPublicKey()
    private var theirSessionKey: ByteArray = KeyPair.createSessionPublicKey()
    private var transcript: ByteArray = KeyExchange.createTranscript()
    private var sendingKey: ByteArray = Message.createSecretKey()
    private var receivingKey: ByteArray = Message.createSecretKey()
    private var sendingNonce: ByteArray = Message.createNonce()
    private var receivingNonce: ByteArray = Message.createNonce()
    private var skippedIndexes: ByteArray = Message.createSkippedIndexes(null)

    companion object {
        init {
            System.loadLibrary("autograph")
        }

        private external fun autographUseKeyPairs(
            identityKeyPair: ByteArray,
            sessionKeyPair: ByteArray,
            ourIdentityKeyPair: ByteArray,
            ourSessionKeyPair: ByteArray,
        )

        private external fun autographUsePublicKeys(
            identityKey: ByteArray,
            sessionKey: ByteArray,
            theirIdentityKey: ByteArray,
            theirSessionKey: ByteArray,
        )
    }

    fun useKeyPairs(
        ourIdentityKeyPair: ByteArray,
        ourSessionKeyPair: ByteArray,
    ): Pair<ByteArray, ByteArray> {
        autographUseKeyPairs(
            this.ourIdentityKeyPair,
            this.ourSessionKeyPair,
            ourIdentityKeyPair,
            ourSessionKeyPair,
        )
        return KeyPair.getPublicKeys(ourIdentityKeyPair, ourSessionKeyPair)
    }

    fun usePublicKeys(
        theirIdentityKey: ByteArray,
        theirSessionKey: ByteArray,
    ) {
        autographUsePublicKeys(
            this.theirIdentityKey,
            this.theirSessionKey,
            theirIdentityKey,
            theirSessionKey,
        )
    }

    fun authenticate(
        ourId: ByteArray,
        theirId: ByteArray,
    ): ByteArray {
        return Auth.authenticate(ourIdentityKeyPair, ourId, theirIdentityKey, theirId)
    }

    fun certify(data: ByteArray?): ByteArray {
        return Cert.certify(ourIdentityKeyPair, theirIdentityKey, data)
    }

    fun verify(
        certifierIdentityKey: ByteArray,
        signature: ByteArray,
        data: ByteArray?,
    ): Boolean {
        return Cert.verify(theirIdentityKey, certifierIdentityKey, signature, data)
    }

    fun keyExchange(isInitiator: Boolean): ByteArray {
        val (transcript, ourSignature, sendingKey, receivingKey) =
            KeyExchange.keyExchange(
                isInitiator,
                ourIdentityKeyPair,
                ourSessionKeyPair,
                theirIdentityKey,
                theirSessionKey,
            )
        this.transcript = transcript
        this.sendingKey = sendingKey
        this.receivingKey = receivingKey
        return ourSignature
    }

    fun verifyKeyExchange(theirSignature: ByteArray) {
        KeyExchange.verifyKeyExchange(transcript, ourIdentityKeyPair, theirIdentityKey, theirSignature)
    }

    fun encrypt(plaintext: ByteArray): Pair<Int, ByteArray> {
        return Message.encrypt(sendingKey, sendingNonce, plaintext)
    }

    fun decrypt(ciphertext: ByteArray): Pair<Int, ByteArray> {
        return Message.decrypt(receivingKey, receivingNonce, skippedIndexes, ciphertext)
    }
}
